### PR TITLE
Languages config-rewrite

### PR DESCRIPTION
**Pull request by mbettini-ox**

### CODE REVIEW

File: source/settings/core.rst [hunk 1]
```diff
@@ -1483,6 +1479,25 @@ See :ref:`settings` for list of all setting groups.
        language_filters_en = lowercase snowball english-possessive stopwords
      }
 
+
+.. dovecot_core:setting:: language_filter_stopwords_dir
+   :seealso: @fts_tokenization
+   :values: @string
+
+     Path to the directory containing stopword files. Stopword files are
+     looked up in ``”<path>”/stopwords_<lang>.txt``.
```
Comment (sirainen) [inline source/settings/core.rst:1516]: the quotes around `<path>` seem unnecessary. Although maybe the whole `<path>` is unnecessary and it should just be:

```
     Path to the directory containing stopword files. The files are named `stopwords_<lang>.txt`.
```

File: source/settings/core.rst [hunk 2]
```diff
@@ -1468,7 +1468,7 @@ See :ref:`settings` for list of all setting groups.
      It filters “qu'”, “c'”, “d'”, “l'”, “m'”, “n'”, “s'” and “t'”.
 
      Do not use at the same time as ``generic`` tokenizer with
-     ``algorithm=tr29 wb5a=yes``.
+     ``language_tokenizer_generic_algorithm = tr29 language_tokenizer_generic_wb5a = yes``.
```
Comment (sirainen) [inline source/settings/core.rst:1499]: I think this needs something like:

```
     ``language_tokenizer_generic_algorithm = tr29`` and ``language_tokenizer_generic_wb5a = yes``.
```

File: source/settings/core.rst [hunk 3]
```diff
@@ -1569,38 +1549,80 @@ See :ref:`settings` for list of all setting groups.
...
+   :seealso: @language_tokenization
+   :values: @boolean
+   :default: no
+
+   Unicode TR29 rule WB5a setting to the tr29 tokenizer. Splits
+   prefixing contracted words from base word. E.g. “l'homme” → “l”
+   “homme”. Together with a language specific stopword list unnecessary
+   contractions can thus be filtered away. This is disabled by default
+   and only works with the TR29 algorithm. Enable by
+   ``language_tokenizer_generic_algorithm = tr29 language_tokenizer_generic_wb5a = yes``.
```
Comment (sirainen) [inline source/settings/core.rst:1623]: similarly here "and" in the middle

File: source/settings/core.rst [hunk 4]
```diff
@@ -1588,6 +1603,15 @@ See :ref:`settings` for list of all setting groups.
      many Japanese search implementations.
 
 
+.. dovecot_core:setting:: language_token_maxlen
+   :seealso: @language_tokenization
+   :values: @uint
+   :default: see note below
```
Comment (sirainen) [inline source/settings/core.rst:1659]: Use `!<filter/tokenizer-specific>` which is similar as used elsewhere.

File: source/settings/core.rst [hunk 5]
```diff
@@ -1376,6 +1376,34 @@ See :ref:`settings` for list of all setting groups.
    name.
 
 
+.. dovecot_core:setting:: language
+   :seealso: @textcat_config_path;dovecot_core
+   :values: @named_filter
```
Comment (sirainen) [inline source/settings/core.rst:1381]: named_list_filter

File: source/settings/core.rst [hunk 6]
```diff
@@ -1376,6 +1376,34 @@ See :ref:`settings` for list of all setting groups.
...
+   :filter: language
+
+   Defines a language to be used in tokenization.
+
+   At least one language must be specified.
+
+   The language listed first is the default and is used when language
+   recognition fails.
```
Comment (sirainen) [inline source/settings/core.rst:1389]: Hmm. This is actually a bit bad. The order of named list filters isn't very clear in all situations, and I was hoping to eventually not give any guarantees about that. Maybe we need to have language_default=yes|no setting, similar to the namespace_inbox=yes|no setting that namespaces have.

File: source/configuration_manual/fts/tokenization.rst [hunk 7]
```diff
@@ -40,9 +44,52 @@ textcat, see :dovecot_core:ref:`textcat_config_path`.
...
-  plugin {
-    language = en de
+  language en {
   }
+  language de {
+  }
+
+Currently supported languages:
+
+Stemming support indicates whether the ``snowball`` filter can be used.
```
Comment (sirainen) [inline source/configuration_manual/fts/tokenization.rst:54]: Could this link to snowball instead?

File: source/installation_guide/upgrading/from-2.3-to-3.0.rst [hunk 8]
```diff
@@ -189,6 +189,19 @@ Removed features and their replacements
 | ``service_count``                                          | Renamed to :dovecot_core:ref:`service_restart_request_count`.                            |
 |                                                            | The default value is set to ``unlimited``. Value ``0`` is now a configuration error.     |
 +------------------------------------------------------------+------------------------------------------------------------------------------------------+
+| ``fts_language_config``                                    | Renamed to :dovecot_core:ref:`textcat_config_path`.                                      |
++------------------------------------------------------------+------------------------------------------------------------------------------------------+
+| ``fts_languages``                                          | Converted into Renamed to :dovecot_core:ref:`language` blocks {}                         |
```
Comment (sirainen) [inline source/installation_guide/upgrading/from-2.3-to-3.0.rst:194]: extra words: "Converted into Renamed to"

File: source/installation_guide/upgrading/from-2.3-to-3.0.rst [hunk 9]
```diff
@@ -189,6 +189,19 @@ Removed features and their replacements
...
++------------------------------------------------------------+------------------------------------------------------------------------------------------+
+| ``fts_filters``                                            | Split into :dovecot_core:ref:`language_filters`                                          |
+|                                                            | :dovecot_core:ref:`language_filter_normalized_icu_id`                                    |
+|                                                            | :dovecot_core:ref:`language_filter_stopwords_dir`                                        |
+|                                                            | :dovecot_core:ref:`language_token_maxlen`                                                |
++------------------------------------------------------------+------------------------------------------------------------------------------------------+
+| ``fts_tokenizers``                                         | Split into :dovecot_core:ref:`language_tokenizer_generic_algorithm`                      |
```
Comment (sirainen) [inline source/installation_guide/upgrading/from-2.3-to-3.0.rst:201]: Is there no generic language_tokenizers link that can be given, like above you link to language_filters?

File: source/configuration_manual/fts/dovecot.rst [hunk 10]
```diff
@@ -291,22 +291,25 @@ Example configuration using OBOX::
     }
   }
 
-  plugin {
-    # Detected languages. Languages that are not recognized, default to the
-    # first enumerated language, i.e. en.
-    fts_languages = en fr # English and French.
+  # This chain of filters first normalizes and lower cases the text, then
+  #  stems the words and lastly removes stopwords.
```
Comment (sirainen) [inline source/configuration_manual/fts/dovecot.rst:295]: extra space after `#`

File: source/configuration_manual/fts/dovecot.rst [hunk 11]
```diff
@@ -291,22 +291,25 @@ Example configuration using OBOX::
...
-    fts_filters = normalizer-icu snowball stopwords
+  # These tokenizers will preserve addresses as complete search tokens, but
+  # otherwise tokenize the text into "words".
 
+  language_tokenizers = generic email-address
+  language_tokenizer_generic_algorithm = simple
+
+  # Detected languages. Languages that are not recognized, default to the
+  # first enumerated language, i.e. en.
+  # English and French.
```
Comment (sirainen) [inline source/configuration_manual/fts/dovecot.rst:306]: These English and French could be comments below, e.g. `language en { # English`

File: source/configuration_manual/fts/tokenization.rst [hunk 12]
```diff
@@ -24,36 +24,89 @@ The lib-fts tokenization library works in the following way:
...
++---------------+---------------------------------------+----------+-----------+
+| tr            | Turkish                               | Yes      | Yes       |
++---------------+---------------------------------------+----------+-----------+
 
 
 Tokenizers
 ^^^^^^^^^^
 
-List of tokenizers to use.
+List of tokenizers to use. Tokenizers can be language specific
```
Comment (sirainen) [inline source/configuration_manual/fts/tokenization.rst:98]: missing period

File: source/settings/core.rst [hunk 13]
```diff
@@ -1376,6 +1376,292 @@ See :ref:`settings` for list of all setting groups.
...
+
+
+.. dovecot_core:setting:: language_filters
+   :seealso: @language_tokenization
+   :values: @boollist
+
+   The list of filters to apply.
+
+   Language specific filter chains can be specified with ``language_filters_<lang>``
+   (e.g. ``language_filters_en``).
```
Comment (sirainen) [inline source/settings/core.rst:1414]: This is now wrong/obsolete

File: source/settings/core.rst [hunk 14]
```diff
@@ -1376,6 +1376,292 @@ See :ref:`settings` for list of all setting groups.
...
+
+.. dovecot_core:setting:: language_tokenizers
+   :default: generic email-address
+   :seealso: @language_tokenization
+   :values: @boollist
+
+   The list of tokenizers to use.
+
+   This setting can be overridden for specific languages by using
+   ``language_tokenizers_<lang>`` (e.g. ``language_tokenizers_en``).
```
Comment (sirainen) [inline source/settings/core.rst:1540]: This is now wrong/obsolete

**Review by sirainen — CHANGES_REQUESTED** [review]

File: source/configuration_manual/fts/dovecot.rst [hunk 15]
```diff
@@ -291,22 +291,24 @@ Example configuration using OBOX::
...
+  language en { # English - 1st language is also the default for detection.
     # This chain of filters will first lowercase all text, stem the words,
     # remove possessive suffixes, and remove stopwords.
-    fts_filters_en = lowercase snowball english-possessive stopwords
-
-    # These tokenizers will preserve addresses as complete search tokens, but
-    # otherwise tokenize the text into "words".
-    fts_tokenizers = generic email-address
-    fts_tokenizer_generic = algorithm=simple
+    language_filters = lowercase snowball english-possessive stopwords
```
Comment (sirainen) [inline source/configuration_manual/fts/dovecot.rst:310]: Can be just `filters` here. I've generally removed the unnecessary prefix in example configs.

File: source/configuration_manual/fts/tokenization.rst [hunk 16]
```diff
@@ -24,36 +24,90 @@ The lib-fts tokenization library works in the following way:
...
 Example::
 
-    fts_tokenizers = generic email-address
-    fts_tokenizer_generic = algorithm=simple
+  language_tokenizers = generic email-address
+  language_tokenizer_generic_algorithm = simple
 
-See :dovecot_plugin:ref:`fts_tokenizers`
+  language en {
+    language_tokenizers = generic
```
Comment (sirainen) [inline source/configuration_manual/fts/tokenization.rst:107]: just `tokenizers`

File: source/configuration_manual/fts/tokenization.rst [hunk 17]
```diff
@@ -63,11 +117,13 @@ List of filters to apply. Filters can be language specific.
 
 Example::
 
-    fts_filters = normalizer-icu snowball stopwords
-    fts_filters_en = normalizer-icu snowball english-possessive stopwords
+  language_filters = normalizer-icu snowball stopwords
 
+  language en {
+    language_filters = normalizer-icu snowball english-possessive stopwords
```
Comment (sirainen) [inline source/configuration_manual/fts/tokenization.rst:123]: just `filters`

File: source/settings/core.rst [hunk 18]
```diff
@@ -1376,6 +1376,308 @@ See :ref:`settings` for list of all setting groups.
...
+        * ``language_tokenizer_generic_wb5a = yes``
+
+   Example:
+
+   .. code-block:: none
+
+    language_filters = normalizer-icu snowball stopwords
+
+    language en {
+       language_filters = lowercase snowball english-possessive stopwords
```
Comment (sirainen) [inline source/settings/core.rst:1510]: just `filters`

File: source/settings/core.rst [hunk 19]
```diff
@@ -1376,6 +1376,308 @@ See :ref:`settings` for list of all setting groups.
...
+
+   Example:
+
+   .. code-block:: none
+
+    language_tokenizers = generic email-address
+    language_tokenizer_generic_algorithm = simple
+
+    language en {
+      language_tokenizers = generic
```
Comment (sirainen) [inline source/settings/core.rst:1602]: just `tokenizers`

File: source/configuration_manual/fts/dovecot.rst [hunk 20]
```diff
@@ -291,22 +291,24 @@ Example configuration using OBOX::
...
+  # These tokenizers will preserve addresses as complete search tokens, but
+  # otherwise tokenize the text into "words".
 
+  language_tokenizers = generic email-address
+  language_tokenizer_generic_algorithm = simple
+
+  # Detected languages. Languages that are not recognized, default to the
+  # first enumerated language, i.e.
+
+  language en { # English - 1st language is also the default for detection.
```
Comment (sirainen) [inline source/configuration_manual/fts/dovecot.rst:307]: Needs updating for the language_default setting

**Review by sirainen — CHANGES_REQUESTED** [review]

File: source/settings/core.rst [hunk 21]
```diff
@@ -1588,6 +1603,15 @@ See :ref:`settings` for list of all setting groups.
     }
 
 
+.. dovecot_core:setting:: language_token_maxlen
```
[the remark this answers is not in the document]
Comment (mbettini-ox) [reply source/settings/core.rst]: no longer there

File: source/settings/core.rst [hunk 22]
```diff
@@ -1364,6 +1364,349 @@ See :ref:`settings` for list of all setting groups.
...
+
+   ``lowercase``
+
+     Change all text to lower case. Supports UTF8, when compiled with libicu
+     and the library is installed. Otherwise only ASCII characters are
+     lowercased.
+
+     Settings:
+
+       * :dovecot_core:ref:`language_filter_lowercase_token_maxlen`
```
Comment (sirainen) [inline source/settings/core.rst:1428]: no language_filter_lowercase_token_maxlen anymore

**Review by sirainen — CHANGES_REQUESTED** [review]

**Pull request merged**